112_nk_improvements.md

```diff
@@ -0,0 +1,181 @@
+Further we put trust in the registrar to validate this correctly. This does not change anything in our threat model, as we already rely on the registrar to make sure that the EK and AK belong together. 
+
+## Design Details
+
+In the registrar we introduce a new field called `transport_key`. This is only used when mTLS is not used by the agent, as otherwise the key used for the mTLS server side is equal to the NK.
+
+# When mTLS is enabled
+The registration process now looks like this:
+
+1. agent sends the usual information: AK, EK, (EK cert) and mTLS certificate
+2. the registrar responds with MakeCredential Challenge and nonce for TPM2_Certify
+3. agent does ActivateCredential and loads mTLS key pair as NK onto TPM, then runs TPM2_Certify on NK loaded in TPM
+4. agent sends HMAC with the secret and UUID, TPM2_PUBLIC of the NK and the attestation and signature data from TPM2_Certify to registrar
+5. registrar validates HMAC checks that the TPM2_PUBLIC part matches the key in mTLS certificate and checks if the nonce is correct.
+6. agent is now active in registrar
+
+When adding agent to attestation:
+
+1. tenant gets agent information from registrar
+2. tenant uses mTLS certificate to establish a connection with the agent and sends payload and U key
+3. tenant adds agent to verifier
+4. verifier uses mTLS certificate to establish a connection with the agent and sends K key on successful attestation
+
+
+# When mTLS is disables
+The registration process now looks like this:
+
+1. agent sends the usual information: AK, EK, (EK cert) and NK in `transport_key` field (PEM encoded)
+2. the registrar responds with MakeCredential Challenge and nonce for TPM2_Certify
+3. agent does ActivateCredential and loads NK onto TPM, then runs TPM2_Certify on NK loaded in TPM
+4. agent sends HMAC with the secret and UUID, TPM2_PUBLIC of the NK and the attestation and signature data from TPM2_Certify to registrar
+5. registrar validates HMAC checks that the TPM2_PUBLIC part matches the key specified as `transport_key` and checks if the nonce is correct.
+6. agent is now active in registrar
+
+When adding agent to attestation:
+
+1. tenant gets agent information from registrar
+2. tenant encrypts payload and U key with NK and sends it to the agent
+3. tenant adds agent to verifier including the NK
+4. verifier uses the NK to encrypt and then send the K key on successful attestation
+
+### Test Plan
+
+- Will be mostly already covered by the existing end-to-end tests
+- New tests for validating objects via TPM2_Certify will be added where necessary
+
+### Upgrade / Downgrade Strategy
+
+The identity quote endpoint will be replaced by the output as described in the user story. Otherwise this change is transparent for the user. 
+
+### Dependency requirements
+
+No new dependencies are required.
+
+## Drawbacks
+
+None.
+
+## Alternatives
+
+- Eliminate to NK fully (would require for us to force enable mTLS when using payloads)
```

Comment:
In order to reach a point where we can make `mTLS` mandatory, we would need to provide a seamless certificate configuration/bootstrapping/rotation solution within Keylime (not sure what form such solution would take).